### PR TITLE
feat(analytics): query duration/row metrics + trace-correlated slow-query log (#1110 PR2)

### PR DESCRIPTION
**Pull request by KDwevedi**
Server half of #1110 (PR2 of 2) — per the implementation plan in https://github.com/egovernments/Citizen-Complaint-Resolution-System/issues/1110#issuecomment-4987649750 (REVISION 2, esp. R9).

## What

- **Per-query OTEL metrics** — every executed analytics SQL query records `pgr.analytics.query.duration.ms` (histogram) + `pgr.analytics.query.rows` (counter), attributes `kpi_id`/`grain`/`tenant`, via `io.opentelemetry:opentelemetry-api` 1.45.0 over `GlobalOpenTelemetry` (pinned to the SDK bundled by the deployed javaagent v2.11.0). Instrumented at `runOne()` — the single choke point — with the KPI context passed in, so **backend-composed KPIs' SOURCE queries each record their own point** and join the slow-query pool (R9/C6).
- **Trace-correlated slow-query log** — one structured line per `_query` request: `analytics.slow_queries traceId=<id> tenant=<t> total=<n> top=[{name,kpiId,tookMs,rowCount}×3]` (top-3 by `tookMs`; errored entries never execute, so they're excluded; emitted from a `finally` so partial failures still log what ran). Trace id: active span (javaagent + Kong w3c propagation → the browser's per-load trace id) > literal `x-trace-id` header (new optional header on `_query`, fallback only) > `-`.
- **`/packs` additive fields** — `packId` (matched `DashboardPack.id` → FE `layout_id` tag), `persona` (the role that made the pack match: first pack-declared role the caller holds — the server's actual `matchesRoles` decision, R9/C7), `recordCount` (see below).
- **Docs** — `docs/observability/dashboard-metrics-server.md`: scraped metric names, Loki query, Prometheus→Loki→Tempo correlation recipe, `recordCount` semantics. Kept as a separate file to avoid cross-PR conflicts; fold into PR1's `dashboard-metrics.md` once both land.

## `recordCount` semantics (read before reviewing)

**Tenant corpus, NOT the caller's ABAC scope**: `count(*)` on `complaint_facts` with the planner's tenant scope semantics (state level `tenant_id LIKE 'ke%'`, city exact — R9/F7), cached in-memory 5 min per tenant, `null` on error (additive, never fails the response). It feeds the `record_count_tier` metric tag, which needs a shared denominator across personas — deliberately independent of department/citizen/boundary scoping (R9/C9).

## Independence / scope

- **Independent of PR1** (client half): all additions are defensive/optional — the new header is optional, the `/packs` fields are additive, the metrics are a no-op without the javaagent. Mergeable in either order.
- **No compose/config changes needed**: pgr-services already runs the OTEL javaagent with `OTEL_METRICS_EXPORTER: otlp`; the collector→Prometheus scrape path and Promtail→Loki shipping already exist.
- **AC deviation (owner-acked on the issue)**: server metric scope is the `/v2/analytics/*` endpoints that actually feed this dashboard, not PGR `_search`/MDMS/DSS (D14 — those are already visible as javaagent spans).

## Validation

- `mvn package` clean (JDK 17 / Boot 3.2.2, via `maven:3.9-eclipse-temurin-17`).
- Full unit suite green: **84 tests, 0 failures** (17 new: no-agent no-op, top-3 selection incl. `<3`/errored entries, log-line shape + injection sanitizer, trace-id fallback chain, recordCount LIKE-vs-exact, cache hit/expiry/keying, error→null).

## Found in integration (bomet deploy)

- **OTLP unit → exporter name suffix**: the collector's prometheus exporter appends a unit-derived suffix when the OTLP `unit` field is set — the duration histogram scraped as `pgr_analytics_query_duration_ms_milliseconds_*` instead of the documented `pgr_analytics_query_duration_ms_*`. Fixed by dropping the `unit` fields (names carry their unit already) — the identical gotcha #1268 fixed on the client. New SDK-backed test pins the instrument names + empty units; docs note the convention. Suite: **85 tests, 0 failures**.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added OpenTelemetry-backed analytics observability: query duration and returned-row metrics, plus per-request structured slow-query logs (top 3).
  * Added optional trace correlation via the `x-trace-id` header.
  * Enhanced analytics pack responses with `packId`, deterministic `persona`, and `recordCount` (tenant corpus count, null when no matching pack).
  * Added short-term caching for tenant record counts to reduce repeat lookups.
* **Documentation**
  * Documented server metrics, slow-query log format, trace correlation, and updated pack response fields.
* **Tests**
  * Added/expanded tests covering telemetry behavior, trace/tenant normalization, caching semantics, and edge cases.

### CODE REVIEW

File: backend/pgr-services/src/main/java/org/egov/pgr/analytics/AnalyticsMetrics.java [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+        try {
+            Attributes attrs = Attributes.of(
+                    ATTR_KPI_ID, orDash(kpiId),
+                    ATTR_GRAIN, orDash(grain),
+                    ATTR_TENANT, orDash(tenant));
+            queryDuration.record(tookMs, attrs);
+            queryRows.add(rowCount, attrs);
+        } catch (RuntimeException e) {
+            log.debug("analytics metrics record failed (ignored): {}", e.toString());
+        }
```
[the remark this answers is not in the document]
Comment (KDwevedi) [reply backend/pgr-services/src/main/java/org/egov/pgr/analytics/AnalyticsMetrics.java:91]: Fixed in 70bd0d6ab — the exception object is now passed to `log.debug` so the stack trace is preserved.

File: backend/pgr-services/src/main/java/org/egov/pgr/analytics/AnalyticsService.java [hunk 2]
```diff
@@ -464,6 +512,48 @@ private Long asOf(){
...
+                                          Long.class, tenantId + "%")
+                    : jdbc.queryForObject("SELECT count(*) FROM complaint_facts WHERE tenant_id = ?",
+                                          Long.class, tenantId);
+            if (count == null) return null;
+            recordCountCache.put(tenantId, new long[]{count, now + RECORD_COUNT_TTL_MS});
+            return count;
+        } catch (Exception e) {
+            log.debug("recordCount for tenant {} failed (returning null): {}", tenantId, e.toString());
+            return null;
+        }
```
[the remark this answers is not in the document]
Comment (KDwevedi) [reply backend/pgr-services/src/main/java/org/egov/pgr/analytics/AnalyticsService.java:554]: Fixed in 70bd0d6ab — `recordCount` now logs the exception object (stack trace preserved) instead of `e.toString()`.

**Comment by vinothrallapalli-eGov** [comment]
@claude review

File: backend/pgr-services/src/main/java/org/egov/pgr/analytics/AnalyticsController.java [hunk 3]
```diff
@@ -103,6 +106,15 @@ public ResponseEntity<Map<String,Object>> getPacks(@RequestBody Map<String,Objec
...
+            // (first pack-role the caller holds, in the pack's declared role order — same
+            // semantics as DashboardPack.matchesRoles); recordCount -> record_count_tier tag
+            // (tenant corpus on complaint_facts, NOT the caller's ABAC-visible subset).
+            out.put("packId", pack.map(DashboardPack::getId).orElse(null));
+            out.put("persona", pack.map(p -> matchingRole(p, callerRoles)).orElse(null));
+            int stateLen = config.getStateLevelTenantIdLength() == null ? 1 : config.getStateLevelTenantIdLength();
+            out.put("recordCount", service.recordCount(tenantId, stateLen));
             return ResponseEntity.ok(out);
         } catch (IllegalArgumentException e) {
             return ResponseEntity.badRequest().body(error(e));
```
[the remark this answers is not in the document]
Comment (KDwevedi) [reply backend/pgr-services/src/main/java/org/egov/pgr/analytics/AnalyticsController.java:126]: Fixed in ab7d5c9ed — recordCount now takes the same pack-match gate as packId/persona: no matching pack (incl. the anonymous PUBLIC floor) → null, and the count query never runs. AnalyticsControllerPacksTest pins both paths.

File: backend/pgr-services/src/main/java/org/egov/pgr/analytics/AnalyticsMetrics.java [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+        // pgr_analytics_query_duration_ms_milliseconds_*. Omitting the unit keeps the
+        // scraped names exactly pgr_analytics_query_duration_ms_* /
+        // pgr_analytics_query_rows_total (same fix as the client's dashboardMetrics.js, #1268).
+        this.queryDuration = meter.histogramBuilder("pgr.analytics.query.duration.ms")
+                .setDescription("Duration of one executed analytics SQL query (per batch entry / compose source)")
+                .build();
+        this.queryRows = meter.counterBuilder("pgr.analytics.query.rows")
+                .setDescription("Rows returned by executed analytics SQL queries")
+                .build();
+    }
```
[the remark this answers is not in the document]
Comment (KDwevedi) [reply backend/pgr-services/src/main/java/org/egov/pgr/analytics/AnalyticsMetrics.java:68]: Fixed in b51b8bc61 — the metric label is now the state-root tenant (QueryTelemetry.stateRootOf, same stateLevelLen semantics as PrincipalScopeResolver), so fabricated sub-tenant suffixes collapse to their real root and a fabricated root never reaches record() (KPI resolution fails at a root with no published catalog). Full tenantId stays in the sanitized slow-query log line; javadoc/docs updated and tests assert the point attributes via an in-memory SDK meter.

**Review by vinothrallapalli-eGov — APPROVED** [review]
